### PR TITLE
ENH: add `np.printoptions`, a context manager

### LINKED ISSUE #8344: The method to undo numpy set_printoptions
In order to undo what ever print options I have set I need to call a function with a massive number of arguments. Why isn't there an argument like default=False and then if I call it with default=True I get back to the original print options. This doesn't have much of an impact in regular programs but it does have a huge impact in jupyter notebooks! A huge amount of computational physics teaching is shifting to these and it is so frustrating to have to have such an ugly line of code for absolutely no reason when it could be fixed in the simplest way.  

This is taken directly from numpys documentation 
np.set_printoptions(edgeitems=3,infstr='inf',
... linewidth=75, nanstr='nan', precision=8,
... suppress=False, threshold=1000, formatter=None)

### PR DESCRIPTION
**Pull request by ev-br**
Just a syntax sugar over the `get_printoptions/set_printoptions` pair:

```
>>> import numpy as np
>>> with np.printoptions(precision=2):
...     print(np.array([2.0]) / 3)
... 
[0.67]
>>> print(np.array([2.0]) / 3)
[0.66666667]
```

Something like this is probably present in everyone's personal toolbox. One can argue this is where this belongs: personal toolboxes and SO answers (there's at least one, likely more).

The argument for having it in numpy is mostly convenience and ease of introducing to newbees: I can see this used fairly early in teaching, especially non-CS students; the implementation is a bit more magical even if is a mere half dozen lines.

EDIT: closes gh-8344

### CODE REVIEW

**Comment by pv** [comment]
Are printoptions threadlocal or global? Should they be threadlocal?

**Comment by eric-wieser** [comment]
I think this is a duplicate PR, but the original was a little overreaching.

@pv: global, unfortunately

**Comment by xoviat** [comment]
Would this work if two threads used the context manager at the same time?

**Comment by eric-wieser** [comment]
@xoviat: no, but nor does manually calling the get and set like users might currently do. Worth warning against in the docstrings though.

Also whatever happens, we fall afoul of PEP550 not being implemented, when it comes to coroutines

**Comment by mhvk** [comment]
Since I don't think it is possible to set print options locally now, it seems we cannot blame the context manager to be global in scope too. I think this is good to have.

File: numpy/core/arrayprint.py [hunk 1]
```diff
@@ -273,6 +273,40 @@ def get_printoptions():
     return _format_options.copy()
 
 
+@contextlib.contextmanager
+def printoptions(*args, **kwds):
```
Comment (eric-wieser) [inline numpy/core/arrayprint.py]: nit: `**kwargs` is a more common spelling

File: numpy/core/arrayprint.py [hunk 2]
```diff
@@ -273,6 +273,40 @@ def get_printoptions():
...
+    """Context manager for setting print options.
+
+    See `set_printoptions` for the full description of available options.
+
+    Examples
+    --------
+
+    `printoptions` sets print options temporarily, and restores them back
+    at the end of the `with`-block:
```
Comment (eric-wieser) [inline numpy/core/arrayprint.py]: I think this belongs in the main doc section above this heading. Maybe the one-line summary should be

> A context manager that sets print options for the scope of the `with` block, restoring the old options at the end

**Comment by charris** [comment]
i wonder if we should move all the context managers into their own file at some point? Having them scattered about make it a bit trickier to find them.

**Comment by ev-br** [comment]
Ugh, missed #3987 completely. I pushed a version which hopefully addresses review comments, the decision is with you guys :-). 

Making array printing options thread-local is separate, and I'd prefer to defer it to a separate PR.

**Review by eric-wieser — APPROVED** [review]
RST formatting might be a little off, but I don't actually know it well enough to be sure. We can always fix it when the release-notes get cleaned up for release

File: doc/release/1.15.0-notes.rst [hunk 3]
```diff
@@ -15,6 +15,14 @@ New functions
...
   arrays.
 
+* `np.printoptions`, the context manager which sets print options temporarily
+  for the scope of the ``with`` block
+
+```
+    >>> with np.printoptions(precision=2):
+    ...     print(np.array([2.0])) / 3
+    [0.67] 
+```
```
Comment (eric-wieser) [inline doc/release/1.15.0-notes.rst]: I think backticks are markdown-only. Two colons after the word "block" might do the right thing.

Comment (mhvk) [reply doc/release/1.15.0-notes.rst]: That's correct.

Comment (ev-br) [reply doc/release/1.15.0-notes.rst]: Indeed, fixed and force-pushed.

**Comment by charris** [comment]
Thanks Evgeni.

**Pull request merged**